server/serverimpl.go

```diff
@@ -45,6 +45,11 @@ type server struct {
 
 	// The network address Server is listening on. Nil if not started.
 	addr net.Addr
+
+	// To gracefully stop the server.
+	serverCtx    context.Context
+	serverCancel context.CancelFunc
+	connWg       *sync.WaitGroup
```

Comment:
Do we need this new WaitGroup? I think `httpServerServeWg` should be ready after all connections are closed (unless Serve() doesn't wait for connections to close - I haven't checked).

Reply:
`httpServerServeWg` is to handle `httpServer.Serve`'s lifecycle.
But, we need to call all connection's `Close()` function to gracefully shutdown.
So, I think the purpose is different and cannot be shared.